### PR TITLE
Run etcd in container instead of systemd service

### LINKED ISSUE #568: With epel enabled on etcd nodes yum update breaks etcd
Hi,
Just a heads up for the community. If you enabled epel on the systems where etcd is built a yum update will overwrite the kismatic version of etcd with a new version which then then break kismatic etcd.
Kismatic etcd creates two files /usr/bin/etcd_k8s  /usr/bin/etcd_networking which will be overwritten if you do yum updates via cron and you have epel enabled. 
The solution is to either put an exclusion in /etc/yum.conf or disable epel. Just an fyi.

--
Ted

### PR DESCRIPTION
**Pull request by dkoshkin**
Fixes #568, partial work on #724 

As part of this PR, I also removed the KET `v1.3` upgrade code.

### CODE REVIEW

File: ansible/group_vars/etcd-k8s.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # etcd-k8s
 # etcd-install
 etcd_install_dir: /etc/etcd_k8s
-etcd_install_bin_name: etcd_k8s
+etcd_name: etcd_k8s
 etcdctl_install_bin_name: etcdctl
```
Comment (alexbrand) [inline ansible/group_vars/etcd-k8s.yaml]: Should we clean this var up as well?

File: ansible/roles/etcd/tasks/main.yaml [hunk 2]
```diff
@@ -12,47 +12,41 @@
...
+    command: "docker run --net=host --volume=/etc/ssl/certs/:/etc/ssl/certs/ --volume={{etcd_install_dir}}:{{etcd_install_dir}}:ro {{ etcd_img }} /usr/local/bin/etcdctl --endpoint='https://127.0.0.1:{{ etcd_service_client_port }}/' --cert-file={{ etcd_certificates.etcd_client }} --key-file={{ etcd_certificates.etcd_client_key }} --ca-file={{ etcd_certificates.ca }} cluster-health"
     register: result
     until: result|success
     retries: 3
     delay: 5
     when: "{{ etcd_insecure_validate|default('false')|bool == false }}"
 
-  - name: verify insecure {{ etcd_install_bin_name }} cluster health
-    command: "{{ bin_dir }}/{{ etcdctl_install_bin_name }} --endpoint='http://127.0.0.1:{{ etcd_service_client_port }}/' cluster-health"
+  - name: verify insecure {{ etcd_name }} cluster health
```
Comment (alexbrand) [inline ansible/roles/etcd/tasks/main.yaml]: Can we get rid of the `insecure` here?

File: ansible/roles/etcd/templates/etcd.insecure.service [hunk 3]
```diff
@@ -1,26 +1,37 @@
...
 Description=etcd key-value store
 Documentation=https://github.com/coreos/etcd
+After=docker.service
+Requires=docker.service
 
 [Service]
 User=root
-Type=notify
-ExecStart={{ bin_dir }}/{{ etcd_install_bin_name }} \
+ExecStartPre=-{{ bin_dir }}/docker stop {{ etcd_name }}
```
Comment (alexbrand) [inline ansible/roles/etcd/templates/etcd.insecure.service]: What's the `-` in front of these commands?

**Comment by dkoshkin** [comment]
I figured the upgrades were a potential source of failure so did loads of testing for the upgrade from different versions of KET with different and the same `etcd` versions, all went well.

File: ansible/roles/etcd/templates/etcd.insecure.service (hunk 3 above)
Comment (dkoshkin) [inline ansible/roles/etcd/templates/etcd.insecure.service]: Ignore failures, most of the time the docker container will not actually be running so this command will fail.

Comment (alexbrand) [reply ansible/roles/etcd/templates/etcd.insecure.service]: Got it!

File: ansible/roles/etcd/templates/etcd.insecure.service [hunk 4]
```diff
@@ -1,26 +1,37 @@
...
+ExecStartPre=-{{ bin_dir }}/docker rm -f {{ etcd_name }}
+ExecStart={{ bin_dir }}/docker run \
+  -p {{ etcd_service_peer_port }}:{{ etcd_service_peer_port }} \
+  -p {{ etcd_service_client_port }}:{{ etcd_service_client_port }} \
+  --volume={{ etcd_service_data_dir }}:/etcd-data \
+  --volume={{ etcd_install_dir }}:{{ etcd_install_dir }}:ro \
+  --volume=/etc/ssl/certs/:/etc/ssl/certs/ \
```
Comment (alexbrand) [inline ansible/roles/etcd/templates/etcd.insecure.service]: Can this be set to read-only?

File: integration/prepare.go [hunk 5]
```diff
@@ -47,7 +45,7 @@ type nodePrep struct {
 
 var ubuntu1604Prep = nodePrep{
 	CommandsToPrepRepo:         []string{copyKismaticKeyDeb, copyKismaticRepoDeb, updateAptGet},
-	CommandsToInstallEtcd:      []string{installCurlApt, installEtcdApt, installTransitionEtcdApt},
+	CommandsToInstallEtcd:      []string{installCurlApt, installEtcdApt},
```
Comment (alexbrand) [inline integration/prepare.go]: Do we still need to install etcd using packages?

**Comment by alexbrand** [comment]
Great stuff here! The last piece that I just started thinking about is whether we should run etcd in the host network namespace. I can't think of a specific reason at the moment, but I thought I would still bring it up.

**Comment by dkoshkin** [comment]
From the [docs](https://coreos.com/etcd/docs/latest/v2/docker_guide.html) and other deployments I've seen it doesn't.
The one thing that is different with the docker deployment is we lost the `http` endpoint on the `127.0.0.1` IP.

**Comment by alexbrand** [comment]
I think this is ready to go into the feature branch. I am going to kick off a full test run, and merge if everything looks good. 

If we find any outstanding work down the line, we can address in a follow-up issue.

**Comment by alexbrand** [comment]
Most failures are related to the disconnected install, which we don't have an answer for right now.

Merging this in.

**Pull request merged**